pkg/kv/kv.go

```diff
@@ -766,3 +770,87 @@ const (
 	// RCCheckTS stands for 'read consistency read with ts check'.
 	RCCheckTS
 )
+
+// ResourceGroupTagBuilder is used to build the resource group tag for a kv request.
+type ResourceGroupTagBuilder struct {
+	sqlDigest  *parser.Digest
+	planDigest *parser.Digest
+	accessKey  []byte
+}
+
+// NewResourceGroupTagBuilder creates a new ResourceGroupTagBuilder.
+func NewResourceGroupTagBuilder() *ResourceGroupTagBuilder {
+	return &ResourceGroupTagBuilder{}
+}
+
+// SetSQLDigest sets the sql digest for the request.
+func (b *ResourceGroupTagBuilder) SetSQLDigest(digest *parser.Digest) *ResourceGroupTagBuilder {
+	b.sqlDigest = digest
+	return b
+}
+
+// SetPlanDigest sets the plan digest for the request.
+func (b *ResourceGroupTagBuilder) SetPlanDigest(digest *parser.Digest) *ResourceGroupTagBuilder {
+	b.planDigest = digest
+	return b
+}
+
+// BuildProtoTagger sets the access key for the request.
+func (b *ResourceGroupTagBuilder) BuildProtoTagger() tikvrpc.ResourceGroupTagger {
+	return func(req *tikvrpc.Request) {
+		b.Build(req)
+	}
+}
+
+// EncodeTagWithKey encodes the resource group tag, returns the encoded bytes.
+func (b *ResourceGroupTagBuilder) EncodeTagWithKey(key []byte) []byte {
+	tag := &tipb.ResourceGroupTag{}
+	if b.sqlDigest != nil {
+		tag.SqlDigest = b.sqlDigest.Bytes()
+	}
+	if b.planDigest != nil {
+		tag.PlanDigest = b.planDigest.Bytes()
+	}
+	if len(key) > 0 {
+		tag.TableId = decodeTableID(key)
```

Comment:
Do we need to consider any error cases here?

Reply:
If the decoded fails, it will be tagged as 0.

Reply:
I've no idea, do we need to distinguish real '0' with the error '0'?

Reply:
I think no need, and the real `table_id` must be greater than 0, https://github.com/pingcap/tidb/blob/master/pkg/meta/meta.go#L253-L254

Reply:
Got it